### DISCUSSION TITLE
The entire process from getting the raw JSON files from MusicBrainz to importing into Virtuoso

### DISCUSSION THREAD

**Discussion by Yueqiao12Zhang** [DDMAL/linkedmusic-datalake#52] [Announcements]
In another issue, @ahankinson mentioned my procedures for reconciling CSVs in OpenRefine. I uploaded detailed history JSON files that are exported directly from OpenRefine. They contain all the steps I went over to reconcile.

I will also describe my steps:
For all columns with "id", I replaced them with the [https://musicbrainz.org/{entity_type}/{id}] reference. They are direct links to MusicBrainz web pages.
For other columns with "names", "title", "genre names", etc, I reconciled them with the Wikidata reconciliation service. If there is no perfect match, I go to the original page in MusicBrainz and check if there is a Wikidata link since some of them have a different name or cannot be found by the reconciliation service, and I will not reconcile that cell if there's none.
After all the reconciliation procedures, I will add a column beside each reconciled column with all the reconciled Wikidata URLs. 
This is ready to export, after we export, we have to go write the mapper file for the RDF conversion. Copy the header to the relations_mapping_{database name}.json, change it into a JSON format with a single dictionary, each column in the header as a key in the dictionary, and fill their values using Wikidata Property links, Wikidata Instance links, Schema.org links, or MusicBrainz documentation links (preferred from best to worst respectively). 
Then we run the csv2rdf_single_subject.p, we will get an out_rdf.ttl, and this is ready to be imported into Virtuoso. We go to Conductor > Linked Data > Quad Store Upload, select the out_rdf.ttl file, give it a name, check the "create graph explicitly" and upload. We can check if the file is successfully uploaded in the Linked Data > Graphs > Graphs section. If it is there, then we can go to the Linked Data > SPARQL, enter the name we gave to the graph in the Default Graph IRI, and perform SPARQL queries.

_Originally posted by @Yueqiao12Zhang in https://github.com/DDMAL/linkedmusic-datalake/issues/48#issuecomment-2145678856_

**Comment by ahankinson** [comment]
Converted to a discussion because there is no actual issue to be solved here.

**Comment by ahankinson** [comment]
So if you do (on your command-line, assuming you have curl installed):

`$ curl -vvvv -L -H "Accept: text/turtle" "http://www.wikidata.org/entity/Q202996"`

You get the `text/turtle` representation of the Radiohead album, OK Computer. You have to dig through it a bit, but eventually you will see this statement: 

```
s:q202996-8A7E76C9-054D-4A31-B489-D236E386E889 a wikibase:Statement,
		wikibase:BestRank ;
	wikibase:rank wikibase:NormalRank ;
	ps:P436 "b1392450-e666-3926-a536-22c65f834433" ;
	psn:P436 <http://musicbrainz.org/release-group/b1392450-e666-3926-a536-22c65f834433> ;
	pq:P1810 "OK Computer" ;
	pq:P4070 wd:Q33328630 ;
	prov:wasDerivedFrom ref:d6c4d8b9b0a646a2b5c7b998951a3a35e82fdedb .
```

Which is to say that the MusicBrainz ID for this album is already reconciled in Wikidata. 

Similarly, if you go from the other direction: 

`$ curl "https://musicbrainz.org/ws/2/release-group/b1392450-e666-3926-a536-22c65f834433?inc=url-rels"`

There is a `<relation@type="wikidata">` block that gives the Wikidata ID for that release in MusicBrainz.

```
<relation type="wikidata" type-id="b988d08c-5d86-4a57-9557-c83b399e3580">
    <target id="d27dd960-9b18-44a0-9bb4-1fa53f58e33e">https://www.wikidata.org/wiki/Q202996</target>
    <direction>forward</direction>
</relation>
```

It's similar in other MusicBrainz resources as well. You just have to visit the page in MusicBrainz and look on the right-hand side, and see if there is a wikidata link. 

So I'm still not entirely sure where OpenRefine fits in here. The two data sources are already reconciled. Are you trying to load MusicBrainz into Virtuoso via RDF? Or something else?

**Reply by Yueqiao12Zhang** [reply]
Some instances of MusicBrainz data are reconciled and have a Wikidata link in their 'relations' entry referencing that specific instance in Wikidata. I extracted them in the newest push. However, since we are converting them into a flattened CSV, other columns (e.g. 'genres', 'artists') are not reconciled thus no Wikidata is there. We would still need OpenRefine to get the Wikidata links for those entries.
The procedure outline is: getting the raw file, converting it into CSV, converting the CSV into RDF, upload into Virtuoso.

**Reply by ahankinson** [reply]
For artists the wikidata link is definitely there. Look up any bands in musicbrainz and I’ll sure there will be a wikidata link.

Not sure about genre.

**Reply by Yueqiao12Zhang** [reply]
In the data dump, not every band has a wikidata link. For example: https://musicbrainz.org/artist/c7ac345b-057f-4553-9eaf-76c6ce824f96

**Reply by ahankinson** [reply]
That's because it doesn't have a Wikidata page... 

https://www.wikidata.org/w/index.php?fulltext=1&search=Forbidden%20Pigs&title=Special%3ASearch&ns0=1&ns120=1

**Reply by Yueqiao12Zhang** [reply]
Are you saying that I do not have to use OpenRefine in the MusicBrainz database conversion?

**Reply by Yueqiao12Zhang** [reply]
But there are many other attributes in the same CSV that are not referenced by Wikidata. For example, the genre of an artist.

**Reply by dchiller** [reply]
> For example, the genre of an artist.

Yes, these are also often reconciled, see relationships here: 
 - https://musicbrainz.org/genre/489ebed8-1299-4761-ba0b-29d381085f82
 - https://musicbrainz.org/genre/68c81274-5770-4e7b-a4bf-ab0d7d425d99
 - https://musicbrainz.org/genre/1a26306a-b3f1-4185-9f7d-2c4070496c2b

**Reply by dchiller** [reply]
> Are you saying that I do not have to use OpenRefine in the MusicBrainz database conversion?

I think at least we need to make sure that we are extracting whatever has **already** been reconciled. There is no point in re-reconciling.

**Reply by Yueqiao12Zhang** [reply]
Ok, I will add this into the code.

**Reply by Yueqiao12Zhang** [reply]
What I meant is that on a single CSV generated from a single JSON, we only get the Wikidata link for the subject, but none of the other attributes. In this case, if we only convert one single JSON into CSV, we only get the Wikidata for the subject, and we still have to reconcile the others using OpenRefine. For example, we have
artist_id, id_wiki, genre_id, genre_name, genre_wiki
as the header of the CSV, we can extract the information of artist_id, id_wiki, genre_id, genre_name from the JSON, but not the genre_wiki, which we have to go through OpenRefine.

**Comment by fujinaga** [comment]
Forbidden Pigs does has a VIAF ID: http://viaf.org/viaf/144390992

Can find out if there’s any documentation on MusicBrainz site, when they would use Wikidata ID or VIAF ID?
If you can’t find any information, let me know. I’ll ask Alastair Porter, who works at MusicBrainz and is a LinkedMusic team member.

On Jun 5, 2024, at 10:13 PM, Yueqiao Zhang ***@***.***> wrote:



In the data dump, not every band has a wikidata link. For example: https://musicbrainz.org/artist/c7ac345b-057f-4553-9eaf-76c6ce824f96

**Reply by ahankinson** [reply]
Here's what I could find:

https://musicbrainz.org/doc/Wikidata

" A few bots scour MusicBrainz and add Wikidata links to entities, but it can also be done manually."

I think it then enters the [potential edits in for a vote](https://musicbrainz.org/search/edits?auto_edit_filter=&order=desc&negation=0&combinator=and&conditions.0.field=link_type&conditions.0.operator=%3D&conditions.0.args=358&conditions.0.args=352&conditions.0.args=733&conditions.0.args=354&conditions.0.args=594&conditions.0.args=353&conditions.0.args=749&conditions.0.args=351&conditions.1.field=status&conditions.1.operator=%3D&conditions.1.args=1&field=Please+choose+a+condition)

This seems to be the source code for one of the bots that adds Wikidata / VIAF / etc. links

https://github.com/loujine/musicbrainz-scripts/blob/master/mb-edit-create_from_wikidata.user.js

Which I found here, in a list of all the bots:

https://wiki.musicbrainz.org/Guides/Userscripts

**Reply by Yueqiao12Zhang** [reply]
Right now my code can extract the Wikidata information if it exists in the data dumps, but I'm not sure if it still needs to be reconciled once more. In the data dumps, only the subject has the Wikidata resources, but not other objects/attributes of the subject. Should other attributes be reconciled?

**Reply by ahankinson** [reply]
You're going to have to be clearer. What objects / attributes are you talking about, specifically?

**Reply by Yueqiao12Zhang** [reply]
For example, what we will have is a CSV with artists as a subject, a Wikidata link associated with it, and its genre:
artist_id, ... , relations_wiki, ... , genre, ...
There will only be one single Wikidata link referencing to the artist, but there is no Wikidata link for his genre unless we track its id to another database. In this case I think we should reconcile the 'genre' column in OpenRefine.

**Reply by Yueqiao12Zhang** [reply]
The 'recordings' database is not reconciled with Wikidata, but I also think that I should reconcile them with their original song names.

**Comment by fujinaga** [comment]
@Yueqiao12Zhang Let's step back and start from the beginning:
First, find a place to document the process. You can update this [https://github.com/DDMAL/linkedmusic-datalake/blob/main/musicbrainz/musicbrainz_pipeline.md](url) if you'd like or start your own new documentation.
This documentation should include the step-by-step explanation of:

1. Where you download the MusicBrainz datasets. Describe briefly what these datasets are and what each JSON file contains, e.g., you said above that the Artist file contains: artist_id, ... , relations_wiki, ... , genre, .... Provide the full list.
2. Store the datasets in the current repo: [https://github.com/DDMAL/linkedmusic-datalake/tree/main/musicbrainz/jsonld](url)
3. If you are only using portion of a file, for experiment purposes (say, the first 20 records), document how you are doing that and possibly a script that accomplishes that. I'm presuming this one script can be used for different JSON files. Store these file as a subdirectory of [https://github.com/DDMAL/linkedmusic-datalake/tree/main/musicbrainz/jsonld](url)
4. At this point, given the subset of a JSON file (e.g., Artist file), I would like you to write a script to convert the JSON file into a CSV file. This should be as generic and simple as possible, so that we can use this script to convert other JSON files we may get from other music databases. Use ChatGPT and Co-Pilot for this purpose (or for any tasks for that matter). ChatGPT tells me to use **pandas** and **json** libraries then define **json_to_csv** Function. Try to use minimum amount of normalization so that the JSON_to_CSV script can be used generically. If specialized normalization is required, this should be clearly documented as to why this is necessary and to which type of JSON files this is necessary (MusicBrainz specific or otherwise).
5. Now, run your JSON_to_CSV script for the subsets of MusicBrainz JSON files to produce CSV files and store them. At this point, there should be no reconciliations, i.e., no URIs unless they already exist in the JSON files.

Let us know when you have created the CSV files so that we can see them. 
Make sure to document every step above.
If all looks good, the next step would be to manually use OpenRefine to reconcile the CSV file.

One of the main goals of the LinkedMusic project is to create a workflow to integrate as many different music databases as cleanly and simply as possible without relying too much on specialized or custom software so that the workflow would be easy to maintain over several years.

**Reply by ahankinson** [reply]
JSON to CSV is a really hard thing to do generically. The nested relationships are hard to represent in a single tabular form without doing tables within tables, and then you wouldn’t use CSV. And different JSON would produce very different CSV.

Given that the end point is RDF anyway, would it make sense to have the JSON go to three columns CSV: subject, predicate, object? That would be about as generic as you can make it, and trivial to serialize to RDF for import into the virtuoso “lake”. It would still be in a native form (it’s still the Musicbrainz data model) just now in a portable format. 

Or it could just go to RDF directly, but CSV is a good format for reviewing the output.

What are the reconciliation goals? Given that most of the obvious things in MusicBrainz and Wikidata are already reconciled, I’m guessing there would only be a subset of the full JSON/CSV that would need to be reconciled between the two. If the goal is to figure out how reconciliation works, then maybe a different dataset that doesn’t have as many links, like Cantus, would be better to reconcile with Wikidata? That is, given a chant(?), try to figure out if it has an entry in Wikidata?

Maybe a good proof of concept for the whole thing is to see if, given a chant in Cantus, you can find a recording of it in Musicbrainz, with Wikidata providing the intermediary links? And vice versa? You can “fudge” it a bit by choosing a known album in Musicbrainz with a known chant in Cantus, and then see if you can reproduce that link just given a set of data in Virtuoso? (The data would  obviously need to contain the answer, but maybe with enough other data from the various sets that it can be a bit of a challenge…)

Feel free to tell me I’m wrong!

**Reply by fujinaga** [reply]
I think we are trying to find out how difficult this JSON to CSV actually is and I thought Yueqiao was making quite a good progress. But you are right, it might be a good idea to go through the workflow using CantusDB first, although we've done this once before and I believe it went pretty smoothly. 
So, @Yueqiao12Zhang, go through the basic five steps I've I outlined previously but  use the CantusDB data instead of MusicBrainz data first. If needed, Dylan or Lucas can provide you with information on how to get the data.

**Reply by fujinaga** [reply]
Re: the three-column CSV: we could do that if we need to, but I'd like to try the very wide CSV with repeated rows (when there are multiple entires in a row, such as an artist belonging to several genres).

I like the CSV as an intermediate step between RDB and RDF, because we can use OpenRefine to manually see what needs reconciling. Eventually, we may be able to skip the CSV step.

Re: Reconciliation goals: For each database, we are integrating, I'd like to know how much manual reconciliation we need to perform, especially manual ones. MusicBrainz (and RISM?) may be the exceptions where we find most of the entries already reconciled.

I like the proof of concept example. @candlecao can try that once we have the new CantusDB data and find a couple of chant recordings in MusicBrainz, convert their JSON data to RDF (via CSV), and try your query.

@ahankinson I appreciate your comments.

**Reply by Yueqiao12Zhang** [reply]
> I think we are trying to find out how difficult this JSON to CSV actually is and I thought Yueqiao was making quite a good progress. But you are right, it might be a good idea to go through the workflow using CantusDB first, although we've done this once before and I believe it went pretty smoothly. So, @Yueqiao12Zhang, go through the basic five steps I've I outlined previously but use the CantusDB data instead of MusicBrainz data first. If needed, Dylan or Lucas can provide you with information on how to get the data.

My steps are really similar to these five steps. I will have two documents that describe how I convert MusicBrainz JSON to CSV, and CSV to RDF. Both steps are generic, which means that any MusicBrainz JSON can use this script to convert to CSV. This is a one-to-one conversion. The CSV to RDF process is also documented. It is generic that any CSVs can be converted into RDF. Any cells in the CSV that has a valid URI will be added to the output RDF as an RDF URI reference, and all other cells will be added as an RDF Literal. This can be a many-to-one conversion, which can merge as many CSVs as needed into one single RDF. It helps the reconciliation because it merges the intersecting databases into one graph, which will put all the reconciled Wikidata references into this graph.

Since all major entities are reconciled, then we can skip all the steps that involve OpenRefine. However, the vertical CSV expansion approach is still used since it makes the most sense to me.

Cantus DB reconciliation, conversion and upload were already done. The current script can still work on the reconciled Cantus DB CSV correctly.

**Reply by fujinaga** [reply]
@Yueqiao12Zhang Can you go through the five steps using a new set of CantusDB? I think the original dump had 50 records, can you go through the workflow with 100 records, or double the original number of records? Please fully document the process so that someone can replicate the process.

**Reply by Yueqiao12Zhang** [reply]
Ok. I will try.

**Reply by Yueqiao12Zhang** [reply]
> @Yueqiao12Zhang Can you go through the five steps using a new set of CantusDB? I think the original dump had 50 records, can you go through the workflow with 100 records, or double the original number of records? Please fully document the process so that someone can replicate the process.

I have worked on a dump with over a thousand records from previous codes: cantusdb/cantusdb_01102024.csv, and that worked fine. I reconciled it using OpenRefine, then I converted it using my script. Would you like for me to work on the large SQL dumps in Cantus DB?

**Reply by fujinaga** [reply]
I would like you to try the entire process using a new set from CantusDB, you can use, say, 2,000 records. Please create a new issue for this and document carefully the process. Once you've completed that I would like you try other databases, e.g., SIMSSA DB and The Session. We are trying to figure how much we can generalize without writing separate codes for different databases. The less code means easier maintenance and easier to scale up.

**Reply by Yueqiao12Zhang** [reply]
Ok. Dylan told me that the CSV dumps for CantusDB will soon be available, and only the SQL dumps are available right now. If we can get the CSV dumps, then no additional code will be needed. But if we use the SQL dumps, we still need to convert SQL to CSV. Should I use the SQL dumps directly?

**Reply by fujinaga** [reply]
No. Wait until CSV is available. Also, use 2,000 records not 7,000. Move onto SIMSSA DB and The Session.

**Comment by ahankinson** [comment]
> Re: the three-column CSV: we could do that if we need to, but I'd like to try the very wide CSV with repeated rows (when there are multiple entires in a row, such as an artist belonging to several genres).

> I like the CSV as an intermediate step between RDB and RDF, because we can use OpenRefine to manually see what needs reconciling. Eventually, we may be able to skip the CSV step.

In my experience, OpenRefine works best when you only have enough data from the dataset you're trying to reconcile to make the link. OpenRefine will try to reconcile *every* row in your CSV. So if you have a set of repeated rows for, say, multiple genres, and then another set of the same repeated rows where each genre is repeated by a different track in each row, then it will try to reconcile every one of the repeated rows. You can choose to reconcile every matching value the same way, but if you're repeating each row for every combination of lists, then there is a lot of duplication in your CSV. You may end up with 20-30 rows (or more) for each genre entry, for each MusicBrainz entity.

So if you have a very simple example like this:

```json
{
"id": "musicbrainz/release/12345"
"name": "A great Release",
"genres": [
    {"name": "pop", "id": "musicbrainz/genre/54321" }, 
    {"name": "rock", "id": "musicbrainz/genre/987654"}
],
"external_links": [
    {"name": "wikidata", "url": "wikidata/Q777777"}, 
    {"name": "discogs", "url": "discogs/33333"}
],
"tracks": [
    {"name": "intro", "id": "musicbrainz/track/666666"}, 
    {"name": "track 2", "id": "musicbrainz/track/55555"},
]}
```

You could expect that in "Wide" CSV format this would serialize as (ignoring the "url/id" in each entry in the list for now):


| id | name | genre | external | track |
|---|------|-------|----------|------|
|musicbrainz/release/12345|A great release|pop|wikidata|intro|
|musicbrainz/release/12345|A great release|rock|wikidata|intro|
|musicbrainz/release/12345|A great release|pop|discogs|intro|
|musicbrainz/release/12345|A great release|rock|discogs|intro|
|musicbrainz/release/12345|A great release|pop|wikidata|track2|
|musicbrainz/release/12345|A great release|rock|wikidata|track2|
|musicbrainz/release/12345|A great release|pop|discogs|track2|
|musicbrainz/release/12345|A great release|rock|discogs|track2|

That's a lot of duplicated entries in the "genre" field when all you really need in OpenRefine is this:

|genre_id|genre|wikidata_id
|---|---|---|
|musicbrainz/genre/54321|pop|?|
|musicbrainz/genre/987654|rock|?|

When I reconciled DIAMM and RISM sources, all I used from DIAMM was the DIAMM id, library siglum and the shelfmark. I created a column that combined the siglum and shelfmark, and used that as the "query" in the OpenRefine service in RISM. When there is a match, you end up with a link between the DIAMM ID (in the sheet) and the RISM ID (from the reconciliation.) You then store the link. 

So It's better to reconcile each genre separately, which gives you a MusicBrainz Genre ID, and the Wikidata Q-id as the reconciled values. You can then either choose to keep those and make the link in your data dump when you process, say, the recordings, OR you could load that directly into your graph database, e.g., add a bunch of statements that essentially say:

`<MusicBrainz-URI> <sameAs> <Wikidata-URI>`

Then when you load your MusicBrainz data into your graph database with all its MusicBrainz URIs (including the Genre ones), it will be pre-populated with the relationships and the Wikidata URIs already, and you will already be able to query those relationships.

**Reply by Yueqiao12Zhang** [reply]
> You could expect that in "Wide" CSV format this would serialize as (ignoring the "url/id" in each entry in the list for now):
> 
> id	name	genre	external	track
> musicbrainz/release/12345	A great release	pop	wikidata	intro
> musicbrainz/release/12345	A great release	rock	wikidata	intro
> musicbrainz/release/12345	A great release	pop	discogs	intro
> musicbrainz/release/12345	A great release	rock	discogs	intro
> musicbrainz/release/12345	A great release	pop	wikidata	track2
> musicbrainz/release/12345	A great release	rock	wikidata	track2
> musicbrainz/release/12345	A great release	pop	discogs	track2
> musicbrainz/release/12345	A great release	rock	discogs	track2

The convert to CSV script eliminates duplicates and only adds necessary information in additional lines. It goes like:
```
 id	name	genre	external	track
 musicbrainz/release/12345	A great release	pop	wikidata	intro
 musicbrainz/release/12345			rock	discogs		track2
```
And when we convert it into RDF and import it into the graph database, there will be 7 triples in total from this CSV.
